src/logged/mod.rs

```diff
@@ -0,0 +1,511 @@
+        ProducerRecord,
+        oneshot::Sender<Result<ProducedOffset, ProducerError>>,
+    )>,
+    root_path: PathBuf,
+}
+
+#[derive(Clone, Deserialize, Debug, Eq, PartialEq, Serialize)]
+struct StorableRecord {
+    version: u32,
+    headers: Option<Vec<Header>>,
+    timestamp: Option<DateTime<Utc>>,
+    key: u64,
+    #[serde(with = "base64_serde")]
+    value: Vec<u8>,
+    offset: u64,
+}
+
+impl FileLog {
+    pub fn new(root_path: PathBuf) -> Self {
+        let retained_root_path = root_path.clone();
+
+        let (producer_tx, mut producer_rx) = mpsc::channel::<(
+            ProducerRecord,
+            oneshot::Sender<Result<ProducedOffset, ProducerError>>,
+        )>(PRODUCER_QUEUE_SIZE);
+
+        tokio::spawn({
+            async move {
+                let mut topic_files: HashMap<Topic, File> = HashMap::new();
+                while let Some((record, reply_to)) = producer_rx.recv().await {
+                    let mut result = Err(ProducerError::CannotProduce);
+
+                    if let Ok(offsets) = find_offset(&root_path, &record.topic).await {
+                        let next_offset = if let Some(offsets) = offsets {
+                            offsets.end_offset.wrapping_add(1)
+                        } else {
+                            0
+                        };
+
+                        let topic_file: Option<&mut File> = if let Some(topic_file) =
+                            topic_files.get_mut(&record.topic)
+                        {
+                            Some(topic_file)
+                        } else {
+                            let topic_file = fs::OpenOptions::new()
+                                .append(true)
+                                .create(true)
+                                .open(root_path.join(&record.topic))
+                                .await;
+                            if let Ok(topic_file) = topic_file {
+                                topic_files.insert(record.topic.clone(), topic_file);
+                                topic_files.get_mut(&record.topic)
+                            } else {
+                                warn!("Error when trying to append to a file: {:?}", topic_file);
+                                increment_counter!("producer_append_failures");
+                                None
+                            }
+                        };
+
+                        if let Some(topic_file) = topic_file {
+                            let storable_record = StorableRecord {
+                                version: 0,
+                                headers: record.headers,
+                                timestamp: record.timestamp,
+                                key: record.key,
+                                value: record.value,
+                                offset: next_offset,
+                            };
+                            if let Ok(buf) = postcard::to_stdvec(&storable_record) {
```

Comment:
No error handling needed? If so, please add comment to explain.

Reply:
There’s a var named result which has been preloaded with an error value.

Reply:
Ah, OK, I missed that. Personally I do not like that implementation pattern too much. Would be good to add a comment before the line initializing the result explaining that below only on the happy path the proper result is built.

Reply:
How might you go about writing this? I’m thinking that the alternative are a lot of else statements. Perhaps when we can concat if let with other expressions (coming), this will be unnecessary.